apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1360,10 +1428,231 @@ private void StartPush()
+            /// This method is called when the acknowledgement request completes.
+            /// This method handles the response from an acknowledgement request for exactly once subscriptions only.
+            /// For non-exactly once subscriptions, <see cref="HandleAckResponse(Task, List{string}, List{string}, List{TimedId})"/> is called.
+            /// </summary>
+            /// <param name="writeTask">The task containing the response of Acknowledge/ModifyAckDeadline RPC.</param>
+            /// <param name="ackIds">The list of ids that were sent to the Acknowledge RPC for acknowledgement
+            /// that the message associated with these ids shouldn't be redelivered.</param>
+            /// <param name="nackIds">The list of ids that were sent to the ModifyAckDeadline RPC to indicate that
+            /// the message associated with these ids should be redelivered.</param>
+            /// <param name="extendIds">The list of ids that were sent to the ModifyAckDeadline RPC to modify their ack deadline.</param>
+            /// <remarks>
+            /// Only one of ackIds, nackIds and extendIds will be non-null and based on that we determine if we are handling acks, nacks or extends.
+            /// This method is used for exactly once subscriptions only. To keep the logic of exactly once subscription in one place,
+            /// this method has a few local methods that are used only in exactly once flow and not in non-exactly once flow, so the method
+            /// is pretty long.
+            /// </remarks>
+            private void HandleAckResponseForExactlyOnce(Task writeTask, List<string> ackIds, List<string> nackIds, List<TimedId> extendIds)
+            {
+                _concurrentPushCount -= 1;
+                _pushInFlight -= (ackIds?.Count ?? 0) + (nackIds?.Count ?? 0) + (extendIds?.Count ?? 0);
+                
+                bool hasAckIds = ackIds?.Count > 0;
+                bool hasNackIds = nackIds?.Count > 0;
+                bool hasExtendIds = extendIds?.Count > 0;
+                // In one call, we'll have one of acks or nacks or extends, not all.
+                var ids = hasAckIds ? ackIds
+                    : hasNackIds ? nackIds : extendIds.Select(j => j.Id);
+
+                if (writeTask.IsFaulted)
+                {
+                    // There may be few successes and few failures or all failures.
+                    var rpcException = writeTask.Exception.As<RpcException>();
+                    if (rpcException != null)
+                    {
+                        var ackError = AckError.ForRpcException(rpcException, ids);
+                        // Only temporary failures should be retried. Mark them as retryable if not already.
+                        UpdateRetryableIds(ids, ackError);
+
+                        if (hasAckIds)
+                        {
+                            // ack = true implies acks.
+                            RetryAcksAndNacks(true, ackError);
```

Comment:
It's best to have two methods `RetryAcks` and `RetryNacks` and then call into `RetryAcksAndNacks` in those methods. Then you'd not need to expose bools and explain them via a comment.